### PR TITLE
Recover anchors on startup

### PR DESCRIPTION
**Pull request by ukstv**
BTW, shaves 5-20 ms from `POST /documents` request as tested locally.

Replaces #860

### CODE REVIEW

File: packages/core/src/anchor/ethereum/ethereum-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-service.ts:177]: we've been using the convention of a leading underscore for private method names, can you put that back?

File: packages/core/src/anchor/ethereum/ethereum-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-service.ts:193]: Can remove the `else` statement and remove a layer of indentation

File: packages/core/src/anchor/ethereum/ethereum-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-service.ts:195]: hmm, I wonder if we should be handling all failed statuses here, not just `HTTP_STATUS_NOT_FOUND`...

Probably out of scope for this change though

File: packages/core/src/anchor/ethereum/ethereum-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/ethereum/ethereum-anchor-service.ts:210]: This is already a large and complex review.  Can we leave formatting changes out of it and do them in a dedicated review if needed?  As-is there's lots of noise in this review from tabbing changes that aren't relevant to the core change.  I don't really have strong feelings about number of spaces to indent with anyway, so I would rather just leave it as-is until we adapt an autoformatter to handle formatting for us.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/memory/in-memory-anchor-service.ts:32]: whooops, I thought I'd done that already.  Nice catch

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/memory/in-memory-anchor-service.ts:58]: So as best I can tell `Subject` is the same as `Observable` but allows multiple subscribers instead of just one.  Why do we need to support multiple subscribers here?

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/memory/in-memory-anchor-service.ts:105]: Nice improvement!

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/memory/in-memory-anchor-service.ts:169]: So I had read somewhere when I was first teaching myself about javascript/typescript a few months ago that it's a bad idea to mix sync and async calls.  I don't totally remember why it was a problem though, something about how the node event loops works.  Do you know what I'm referring to?  Is this something I shouldn't actually be worried about?

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [inline packages/core/src/anchor/memory/in-memory-anchor-service.ts:239]: why do the anchor requests need to get serialized here?  What's wrong with letting them happen in parallel like they are currently?

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:58]: Okay, I see now, it's because we send all the anchor requests through this single feed, then filter the global feed to individual per-request observables that we return from `requestAnchor`.  So then why have this single unified `#feed` at all?  The ethereum-anchor-service doesn't do this - why can't the in-memory anchor service do the same and just have a bunch of totally separate per-request observables without ever serializing them into a single feed?

File: packages/core/src/ceramic.ts
Comment (stbrody) [inline packages/core/src/ceramic.ts:75]: I'm a little sad to be adding testing-only configuration options here.  Not the end of the world certainly, but would be nice to avoid if we can do so without causing too much pain in our tests.  I'm actually kind of surprised that this causes issues in the tests - I would have expected the `restoreDocuments()` call to be a no-op because it never actually finds any pinned documents that need restoring.

**Review by stbrody — COMMENTED** [review]
Overall this looks very nice.  A bunch of little comments, mostly style nitpicks and questions to firm up my understanding of rxjs, but nothing major.  I'm definitely starting to see the value of rxjs more as I understand this review better - the improved error handling story alone is pretty motivating.

My biggest concern remains whether this is the right time to roll out rxjs.  I'm a little hesitant to put this in without pretty quickly thereafter following it up with removing EventEmitter and replacing it with rxjs throughout the entire codebase.  Otherwise there's two mental models for handling events and synchronization that developers need to get familiar with to operate in our code.  So I wonder if we decide to go in this direction if we should put together a project soon to fully switch over.  If not, then do we want to hold off on putting rxjs *anywhere* until we can roll it out throughout the codebase?  Would be interested in your thoughts here @oed

File: packages/common/src/anchor-service.ts
Comment (ukstv) [reply packages/common/src/anchor-service.ts:40]: Type-safety. Less opportunity to shoot yourself in the foot. Here it allows one to... Whoops, it does not, yet.

Now it does with one more commit.

From just one single ASR you could not figure what fields are expected to be populated for what status. Source of confusion, source of errors, source of delays. It too quickly deteriorates into stringly-typed programming. With this change we have total clarity regarding fields, and compiler [helps to remind](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#union-exhaustiveness-checking) if some of the variants one forgets to handle.

File: packages/core/src/anchor/ethereum/ethereum-anchor-service.ts
Comment (oed) [reply packages/core/src/anchor/ethereum/ethereum-anchor-service.ts:210]: Just fyi, the autoformatter (pretty) we use in other repos enforce 2 spaces indention.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (oed) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:169]: Why is that relevant here? This function is completely sync so it should not be async.

**Review by oed — COMMENTED** [review]
This looks mostly good to me :) 

> My biggest concern remains whether this is the right time to roll out rxjs. I'm a little hesitant to put this in without pretty quickly thereafter following it up with removing EventEmitter and replacing it with rxjs throughout the entire codebase. Otherwise there's two mental models for handling events and synchronization that developers need to get familiar with to operate in our code.

@stbrody I don't feel super strongly that this needs to happen fast. However we might make a large chunk of it part of the state refactor?

File: packages/core/src/__tests__/ceramic-recover-documents.test.ts
Comment (ukstv) [reply packages/core/src/__tests__/ceramic-recover-documents.test.ts:150]: Good catch. I guess, for this purpose, InMemoryAnchor service adds a timeout when anchoring. Anyway, the whole `ceramic.test.ts` is to be fixed at least.

Comment (stbrody) [reply packages/core/src/__tests__/ceramic-recover-documents.test.ts:150]: Yep that's right.  `ceramic.test.ts` hasn't been updated yet since we've started switching to the new paradigm to prevent races

File: packages/core/src/document.ts
Comment (ukstv) [reply packages/core/src/document.ts:597]: Unlike `mergeMap`, concatMap stops the feed propagation further until the function inside `concatMap` is done.
`interval` inside EthereumAnchorService pushes events every N seconds. I would like to be sure, that processing of those happens sequentially, instead of having multiple request-processing instances at a time.

File: packages/core/src/subscription-set.ts
Comment (ukstv) [reply packages/core/src/subscription-set.ts:15]: I was as baffled as you when I read it first.

File: packages/core/src/anchor/ethereum/ethereum-anchor-service.ts
Comment (ukstv) [reply packages/core/src/anchor/ethereum/ethereum-anchor-service.ts:177]: Okay, but this convention makes me sad, like Hungarian notation.

File: packages/core/src/anchor/ethereum/ethereum-anchor-service.ts
Comment (ukstv) [reply packages/core/src/anchor/ethereum/ethereum-anchor-service.ts:210]: #879

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (ukstv) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:58]: The logic here is different. Unlike independent requests for ethereuem, all of these requests are triggered by  single call to `#anchor`. That's why.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (ukstv) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:169]: https://nodejs.org/en/docs/guides/blocking-vs-non-blocking/ - irrelevant here. Nothing here relies on event-loop.

File: packages/core/src/ceramic.ts
Comment (ukstv) [reply packages/core/src/ceramic.ts:75]: You would be surprised to know, that the whole Inmemory anchor service is just for test purpose only.

Comment (ukstv) [reply packages/core/src/ceramic.ts:75]: This is the least intrusive option to avoid re-mocking tests that apparently heavily rely on state store being fully mocked.

Comment (ukstv) [reply packages/core/src/ceramic.ts:75]: And, honestly, if the initialisation of Ceramic is done differently, this should go away.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:58]: Ah right, that makes sense.  The in-memory anchor service supports `anchorOnRequest:false` which the eth anchor service does not.

File: packages/core/src/document.ts
Comment (stbrody) [inline packages/core/src/document.ts:598]: Can we get rid of this now that we're getting an observable per anchor request instead of a single event feed from the entire anchor service?  The Observable returned from `AnchorService.requestAnchor` should be able to only emit events that are responses to the specific anchor request right?

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:169]: Yeah, that must have been what I was remembering.  Guess I was overthinking it.  Thanks!

File: packages/core/src/ceramic.ts
Comment (stbrody) [reply packages/core/src/ceramic.ts:75]: > You would be surprised to know, that the whole Inmemory anchor service is for test purpose only.

This is the first InMemoryAnchorService-specific config option to show up in this `CeramicConfig` interface, that's why I'm calling it out.

Comment (stbrody) [reply packages/core/src/ceramic.ts:75]: > This is the first InMemoryAnchorService-specific config option to show up in this CeramicConfig interface, that's why I'm calling it out.

I guess there's also `anchorOnCommit`, which doesn't actually show up explicitly here in CeramicConfig, but is implicitly a member as it is passed to the InMemoryAnchorService via this object because `CeramicConfig` allows arbitrary extra options.  I don't really like that either, but maybe doing something similar would be better than polluting this highly user-visible configuration interface with testing-only parameters

Comment (stbrody) [reply packages/core/src/ceramic.ts:75]: > This is the least intrusive option to avoid re-mocking tests that apparently heavily rely on state store being fully mocked.

So the issue is that many of the unit tests don't have a real state store, so in order to get the tests to pass without this option, you'd need to update the mocked state store in those tests to expect the initial call to `stateStore.list()` and return an empty list?

Assuming that is indeed the issue, then I think it's fair to want to avoid doing that in a large number of our tests.  I think in that case the thing to do would be to remove this argument from `CeramicConfig` and handle it similarly to how `anchorOnCommit` is handled.  Would be nice to also set it to default to false if we're running in testing mode, similar to how the `useCentralizedPeerDiscovery` default is set

Comment (stbrody) [reply packages/core/src/ceramic.ts:75]: > And, honestly, if the initialisation of Ceramic is done differently, this should go away.

How would that work?

File: packages/core/src/document.ts
Comment (stbrody) [reply packages/core/src/document.ts:597]: interesting.  So `pipe` doesn't process the events sequentially already?  It just seems weird to me since the docs for `concatMap` all talk about merging multiple input observables into one output observable, but here we only have 1 input observable.  Is there no more specific and semantically meaningful operator to take a single observable and just tell it to finish processing one event before starting to process the next?

**Review by stbrody — COMMENTED** [review]
Okay, I think I responded to all the comment threads that you responded to, though there are still at least a few comments from my original pass that didn't get a response yet

File: packages/core/src/document.ts
Comment (ukstv) [reply packages/core/src/document.ts]: True.

**Comment by ukstv** [comment]
"Is there no more specific and semantically meaningful operator to take a single observable and just tell it to finish processing one event before starting to process the next" Please, go find and propose one, if you feel like there is a better way.

**Comment by ukstv** [comment]
> S.U.: And, honestly, if the initialisation of Ceramic is done differently, this should go away
> S.B.: How would that work?

As I told you before, separate parameter configuration based on input params from actual initialisation. You have external parameters, internal configuration, and components initialisation as three separate entities.

> but maybe doing something similar would be better than polluting

Spencer, please! Gorilla InMemoryAnchorService exists just for testing purposes, yet it is a full-fledged part of the production code. InMemoryAnchorService's totally opaque parameters, that are not declared anywhere, are casted from `config as any`. Tests call `.anchor` on it after telling compiler, "please, close your eyes, here is a dirty thing going". _This_ is pollution.

Some day someone does proper initialisation. For now, we have both `restoreDocuments` and `useCentralizedPeerDiscovery` as part of config. Oops, mind doing `git blame` for the latter?

**Comment by oed** [comment]
Seems like we need to do some initialization cleanup at some point for sure. I'd rather not get stuck on if we should or shouldn't use a param only for testing rn though, unless there is a strong reason for either :)

**Comment by oed** [comment]
> For now, we have both restoreDocuments and useCentralizedPeerDiscovery as part of config. Oops, mind doing git blame for the latter?

Hey! Let's try to not point fingers here. Seems like Spencer is trying to shoot for the ideal, and maybe that's not the right thing here now, but please try to be more friendly when interacting other people :)

**Comment by stbrody** [comment]
> "Is there no more specific and semantically meaningful operator to take a single observable and just tell it to finish processing one event before starting to process the next" Please, go find and propose one, if you feel like there is a better way.

I'm asking because I'm not familiar at all with rxjs, so I was wondering if you knew more about the available APIs.  So I guess you're saying that the standard way in rxjs to make an event stream fully process one event before beginning to process the next is to always use `mergeMap`, even if there's actually only one stream and therefore nothing to "merge"?

**Comment by stbrody** [comment]
> Gorilla InMemoryAnchorService exists just for testing purposes, yet it is a full-fledged part of the production code. InMemoryAnchorService's totally opaque parameters, that are not declared anywhere, are casted from `config as any`. Tests call `.anchor` on it after telling compiler, "please, close your eyes, here is a dirty thing going". _This_ is pollution.

It's true that a bunch of things about the `InMemoryAnchorService` are ugly, especially the part that it ships as part of our production code when it is only intended for testing, but the difference here is that none of the config options for `InMemoryAnchorService` are part of our publicly documented APIs.  `restoreDocuments` would be the first.  I'm fine with a certain amount of ugliness under the covers, but generally try to be more careful with what shows up in our user-visible APIs.

> Some day someone does proper initialisation. For now, we have both `restoreDocuments` and `useCentralizedPeerDiscovery` as part of config.

`useCentralizedPeerDiscovery` controls public behavior, it makes sense for it to be part of the public configuration API.  We change the default based on whether or not we're in testing mode, but the actual behavior being configured is relevant outside of tests.

**Comment by oed** [comment]
Is there any blockers here besides bike shedding around constructor arguments?

**Comment by ukstv** [comment]
> I'm asking because I'm not familiar at all with rxjs, so I was wondering if you knew more about the available APIs. So I guess you're saying that the standard way in rxjs to make an event stream fully process one event before beginning to process the next is to always use mergeMap, even if there's actually only one stream and therefore nothing to "merge"?

A `Promise` is an `Observable`, semantically. The difference is Observable is a feed of generally unknown length,  while Promise is guaranteed to emit one value. So, `concatMap`/`mergeMap` actually merges two observables, even if Observable nature of Promise is hidden behind technicalities.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (stbrody) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts:239]: I'm still curious about this for my own understanding in case I'm missing something, but it's not a blocker, especially since this is test-only code

**Review by stbrody — APPROVED** [review]

**Comment by ukstv** [comment]
> I'm still curious about this for my own understanding in case I'm missing something, but it's not a blocker, especially since this is test-only code

Actually, this is a totally legit concern. Fixed. Now anchors go in parallel.

File: packages/core/src/anchor/memory/in-memory-anchor-service.ts
Comment (ukstv) [reply packages/core/src/anchor/memory/in-memory-anchor-service.ts]: Aargh, should be posted here. Valid concern. Fixed.

**Comment by ukstv** [comment]
@oed approval to merge this?

**Review by oed — APPROVED** [review]

**Comment by ukstv** [comment]
Yahoo!

**Pull request merged**
